src/main/java/com/google/devtools/build/lib/packages/SelectorList.java

```diff
@@ -110,7 +110,12 @@ static SelectorList concat(Object x, Object y) throws EvalException {
 
   @Override
   public SelectorList binaryOp(TokenKind op, Object that, boolean thisLeft) throws EvalException {
-    if (op == TokenKind.PLUS) {
+    if (getNativeType(that).equals(Dict.class)) {
```

Comment:
The error message thrown when concatenating the wrong types using the wrong operator (like 146 of SelectorList.java) is now misleading (currently it's hard-coded as "'+' operator applied ..."). To fix, you'd need to pass the operator name from binaryOp to canConcatenate.

Reply:
Ended up moving the validation up to `binaryOp`, since `SelectorList.of` is called in other places where there is not a binary operator available (e.g.  [BuildType.java#L493](https://github.com/bazelbuild/bazel/blob/master/src/main/java/com/google/devtools/build/lib/packages/BuildType.java#L493) and [StarlarkNativeModule.java#L704](https://github.com/bazelbuild/bazel/blob/master/src/main/java/com/google/devtools/build/lib/packages/StarlarkNativeModule.java#L704))